LibVLCSharp.Forms/Shared/MediaPlayerElement.xaml.cs

```diff
@@ -0,0 +1,210 @@
+        {
+            var videoView = VideoView;
+            if (videoView != null)
+            {
+                videoView.MediaPlayer = mediaPlayer;
+            }
+            var playbackControls = PlaybackControls;
+            if (playbackControls != null)
+            {
+                playbackControls.MediaPlayer = mediaPlayer;
+            }
+        }
+
+        private void OnPlayControlsChanged(PlaybackControls playbackControls)
+        {
+            if (playbackControls != null)
+            {
+                playbackControls.MediaPlayer = MediaPlayer;
+                playbackControls.VideoView = VideoView;
+            }
+        }
+
+        private static void VideoViewPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+        {
+            ((MediaPlayerElement)bindable).OnVideoViewChanged((VideoView)newValue);
+        }
+
+        private static void LibVLCPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+        {
+            ((MediaPlayerElement)bindable).OnLibVLCChanged((LibVLC)newValue);
+        }
+
+        private static void MediaPlayerPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+        {
+            ((MediaPlayerElement)bindable).OnMediaPlayerChanged((LibVLCSharp.Shared.MediaPlayer)newValue);
+        }
+
+        private static void PlaybackControlsPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+        {
+            ((MediaPlayerElement)bindable).OnPlayControlsChanged((PlaybackControls)newValue);
+        }
+
+        /// <summary>
+        /// Invoked whenever the <see cref="Element.Parent"/> of an element is set. 
+        /// Implement this method in order to add behavior when the element is added to a parent.
+        /// </summary>
+        /// <remarks>Implementors must call the base method.</remarks>
+        protected override void OnParentSet()
+        {
+            base.OnParentSet();
+
+            if (Parent != null && !Initialized)
+            {
+                Initialized = true;
+
+                if (VideoView == null)
+                {
+                    VideoView = new VideoView();
+                }
+                if (PlaybackControls == null)
+                {
+                    PlaybackControls = new PlaybackControls();
+                }
+
+                var application = Application.Current;
+                application.PageAppearing += PageAppearing;
+                application.PageDisappearing += PageDisappearing;
+            }
+        }
+
+        private void PageAppearing(object sender, Page e)
+        {
+            if (e == this.FindAncestor<Page>())
+            {
+                MessagingCenter.Subscribe<LifecycleMessage>(this, "OnSleep", m =>
+                {
+                    var applicationProperties = Application.Current.Properties;
+                    applicationProperties["VLC_MediaPlayerElement_Position"] = MediaPlayer?.Position;
```

Comment:
Damn, you're right, I did not think about this case. I'm not sure either that's the best way to go. Perhaps, we should let the developper manage the app lifecycle (and just provide practical methods to make it easier for him to recreate the videoview).

Reply:
> Perhaps, we should let the developper manage the app lifecycle (and just provide practical methods to make it easier for him to recreate the videoview).

Yes, maybe. There seems to be no way for us (from the library code) to know when the user's app is going to background. 
The less adjustment the user has to make to his project to get up and running the better. The solution you have for now works well for most cases. 

For more advanced cases (multiple videos, background audio, etc), users will want to handle the lifecycle/playback themselves, and they can (by not firing those messages). I guess we could make those message callbacks public..?

Reply:
@mfkl : Do you suggest something like this?

```cs
var state = mediaElement.SaveState();
// Serialize the state the way you want

// On restore
mediaElement.RestoreState(state);
```
Do you have something else in mind?

Reply:
Something like that. 

If we store the `MediaElementState` in the `Application Properties` we can make the parameter of `RestoreState` optional, so that the user does not have hold of it. But essentially yes.

Reply:
What about saving the state of multiple players? Wouldn't that cause conflicts?

Reply:
I'm not sure we should handle lifecycle events here. The user has the MediaPlayer already, and he can save/restore the state programmatically, as in any other LVS application.

Reply:
> What about saving the state of multiple players? Wouldn't that cause conflicts?

We could use 
```
var positionKey = $"VLC_{MediaPlayer.NativeReference}_MediaPlayerElement_Position";
```
since the MediaPlayer is not disposed when the app is backgrounded. If the user wants to dispose it, then they just don't fire the lifecycle message and handle everything themselves.

> I'm not sure we should handle lifecycle events here.

We just offer to do the basic lifecycle management by offering the user to send a simple message. They are still free to _not_ send the message and handle everything themselves, but we still handle the basic use case gracefully.